src/utils/signals/worker.ts

```diff
@@ -0,0 +1,100 @@
+import { HttpTransportType, HubConnection, HubConnectionBuilder } from '@microsoft/signalr';
+import type { WorkerIncomingMessage, WorkerOutgoingMessage } from './types';
+import { Deferred, EventEmitter } from './utils';
+import { env } from '~/env/client.mjs';
+
+// --------------------------------
+// Types
+// --------------------------------
+interface SharedWorkerGlobalScope {
+  onconnect: (event: MessageEvent) => void;
+}
+const _self: SharedWorkerGlobalScope = self as any;
+
+let connection: HubConnection | undefined;
+const events: Record<string, (data: unknown) => void> = {};
+const deferred = new Deferred<void>();
+
+const emitter = new EventEmitter<{
+  connectionReady: undefined;
+  connectionClosed: { message?: string };
+  connectionError: { message?: string };
+  eventReceived: { target: string; payload: any };
+}>();
+
+const getConnection = async ({ token }: { token: string }) => {
+  if (connection) return connection;
+
+  connection = new HubConnectionBuilder()
+    .withUrl(`${env.NEXT_PUBLIC_SIGNALS_ENDPOINT}/hub`, {
+      accessTokenFactory: () => token,
+      skipNegotiation: true,
+      transport: HttpTransportType.WebSockets,
+    })
+    .withAutomaticReconnect()
+    .build();
+
+  await connection.start();
```

Comment:
I wonder if connection fails to start, would this throw an exception? 🤔 should we be wrapping this in a try/catch so it doesn't break the app? Or because it's running in a shared worker it doesn't matter?